Remove 'dry run' functionality throughout.

jaraco opened the pull request:
<!-- First time contributors: Take a moment to review https://setuptools.pypa.io/en/latest/development/developer-guide.html! -->
<!-- Remove sections if not applicable -->

## Summary of changes

At first, I considered deprecating this functionality. Then I realized that this functionality is exposed almost exclusively through a parameter to `setup.py`, which is already deprecated. Moreover, it seems unlikely that anyone would use this functionality in today's modern usage. I personally haven't used it even going back to the most legacy usage.

Removing these changes from Setuptools should unblock the failing tests in pypa/distutils#335, and in fact, maybe they should be merged together for easier rollback if needed. [Update] The changes from pypa/distutils#335 are included in this merge.

Ref pypa/distutils#334

### Pull Request Checklist
- [x] Changes have tests
- [x] News fragment added in [`newsfragments/`].
  _(See [documentation][PR docs] for details)_


[`newsfragments/`]: https://github.com/pypa/setuptools/tree/master/newsfragments
[PR docs]:
https://setuptools.pypa.io/en/latest/development/developer-guide.html#making-a-pull-request

jaraco committed:
Remove 'dry run' functionality throughout.

jaraco committed:
Remove 'dry run' functionality throughout.

jaraco commented on setuptools/_vendor/wheel/bdist_wheel.py:
```diff
@@ -440,11 +440,10 @@ def run(self):
-                    rmtree(self.bdist_dir, onerror=remove_readonly)
-                else:
-                    rmtree(self.bdist_dir, onexc=remove_readonly_exc)
+            if sys.version_info < (3, 12):
+                rmtree(self.bdist_dir, onerror=remove_readonly)
+            else:
+                rmtree(self.bdist_dir, onexc=remove_readonly_exc)
```
Oops. I notice now this change needs to happen in the vendored wheel project.

jaraco replied:
Reverted in fb1013d58.

jaraco committed:
Add news fragment.

jaraco committed:
Add a dry_run property for temporary compatibility.

Avasam commented on setuptools/__init__.py:
```diff
@@ -167,6 +167,12 @@ class Command(_Command):
     command_consumes_arguments = False
     distribution: Distribution  # override distutils.dist.Distribution with setuptools.dist.Distribution
 
+    dry_run = False
```
stdlib distutils types this using an old `Literal[0, 1]` instead of a boolean.
This'll fix itself with https://github.com/pypa/setuptools/issues/4689
In the mean time you can either type/pyright ignore, or use `0`

jaraco replied:
```suggestion
    dry_run = False  # pyright: ignore # (until #4689; see #4872)
```

Avasam replied:
Quick update: https://github.com/python/typeshed/pull/13615 got merged in typeshed
So next version of mypy that includes a typeshed sync should also resolve this.

jaraco committed:
Remove reliance on --dry-run parameter in test.

jaraco commented:
@abravalheri Any reservations before I proceed with merging?

jaraco committed:
Merge branch 'debt/dry-run' of https://github.com/pypa/distutils into debt/dry-run

jaraco committed:
Merge branch 'main' into debt/dry-run

jaraco commented:
Well that's annoying. The test is failing with:

```
FAILED setuptools/dist.py::ruff - pytest_ruff.RuffError: FURB171 Membership test against single-item container
```

But the test failure report doesn't indicate where the error exists and instead shows a traceback about how it came to report the failure (unhelpful, pytest-ruff).

The test doesn't fail if I run ruff in standalone mode.

I tested on main and the failure isn't here, so I guess I'll bisect the diff and see if I can figure out where the failure is.

jaraco commented:
Oh, I was wrong - the ruff error is visible in the traceback, and the issue is in line 617.

jaraco committed:
Fix FURB171 nitpick by altering the logic.

Removed "ugh!" comment because it's not at all clear if it's even relevant, and even if it was, it's not actionable.

jaraco commented:
> Please update the documentation at https://setuptools.pypa.io/en/stable/deprecated/distutils/apiref.html. I understand that `distutils` is deprecated but it is not unsupported yet, so it should be mentioned that `dry_run` parameter is no longer present.

Thanks for catching this. Docs updated in pypa/distutils@a1c0ebc14, which will merge here in due course.

pkgw commented:
Reporting as requested in the release notes: this change [broke gobject-introspection](https://gitlab.gnome.org/GNOME/gobject-introspection/-/issues/575).
